File: 1. Two Sum/twoSum.py
class Solution(object):
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        cache = {}
        output = None
        for i in range(len(nums)):
            val = nums[i]
            pair = target - val
            if val in cache:
                output = [cache[val], i]
                break
            else:
                cache[pair] = i
        return output

File: 1. Two Sum/TwoSum.java
public class Solution {
    //Use the map strcture in which each pair is consisted of 
    //1) the ind of a certain element in nums & (value);
    //2) the value of the num in this index (key).
    //During building the map, we also check whether there exists a solution,
    //that is, for an index i, find whether map contains key whose value equals to target substract nums[i]. if yes return.
    public int[] twoSum(int[] nums, int target) {
        int[] output = new int[2];
        int counter = 0;
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int num : nums){
            int remainder = target - num;
            if(map.containsKey(remainder)){
                output[0] = map.get(remainder);
                output[1] = counter;
            }
            else{
                map.put(num,counter);
            }
            counter++;
        }
        return output;
    }
}